modules/virt-creating-a-service-from-a-virtual-machine.adoc

```diff
@@ -0,0 +1,196 @@
+                bus: virtio
+          interfaces:
+          - masquerade: {}
+            name: default
+        resources:
+          requests:
+            memory: 1024M
+      networks:
+        - name: default
+          pod: {}
+      volumes:
+        - name: containerdisk
+          containerDisk:
+            image: kubevirt/fedora-cloud-container-disk-demo
+        - name: cloudinitdisk
+          cloudInitNoCloud:
+            userData: |
+              #!/bin/bash
+              echo "fedora" | passwd fedora --stdin
+----
+<1> The `special: key` label
++
+
+. Edit the `Service` YAML to create and expose the Service object, connecting by
+specifying the service name. `Service` is scoped by `namespace`.
+.. Specify the `name` for the service you are creating and exposing.
+.. Add a `namespace` key to the `metadata` section of the Service YAML, that corresponds to the
+`namespace` you referenced in the virtual machine YAML.
+.. In the `spec` section of the Service YAML, add `special: key` to the `selector` attribute,
+which corresponds to the `labels` you added in the virtual machine YAML configuration file.
+.. In the `spec` section of the Service YAML, add `type: ClusterIP` for a
+ClusterIP service. Add `targetPort: 22`, exposing the service on SSH port `22`.
++
+To create and expose other types of services externally, outside of the cluster,
+such as `NodePort` and `LoadBalancer`, replace
+`type: ClusterIP` with `type: NodePort` or `type: LoadBalancer`, as appropriate,
+in the Service object definition.
++
+
+[source,yaml]
+----
+apiVersion: v1
+kind: Service
+metadata:
+  name: vmservice <1>
+  namespace: example-namespace <2>
+spec:
+  ports:
+  - port: 27017
+    protocol: TCP
+    targetPort: 22 <3>
+  selector:
+    special: key <4>
+  type: ClusterIP <5>
+----
+<1> Service name
+<2> `namespace` corresponds to the `namespace` you referenced in the virtual machine YAML.
+<3> Exposes SSH `targetPort` 22
+<4> `selector` `special: key` corresponds to the `labels` you added in the virtual machine YAML.
+<5> Defines Service `type: ClusterIP`
+
++
+. Create the `ClusterIP` service using `kubectl`.
++
+
+[source,terminal]
+----
+$ kubectl create -f <service_name>.yaml
+----
+
++
+. Start the virtual machine. If the virtual machine is already running, restart it.
++
+
++
+. Query the Service object to verify it is available and is configured with type `ClusterIP`.
++
+
+[source, terminal]
+----
+$ kubectl get service
+----
++
+
+.Example output
+[source, terminal]
+----
+NAME        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)     AGE
+vmservice  ClusterIP   172.30.3.149   <none>        27017/TCP   2m
+----
++
+
+. Connect to the `VirtualMachine` using the service IP and service port.
+* Exposing a virtual machine's SSH port through a `ClusterIP` service enables
+connection to a virtual machine within the cluster, either through
+another pod or virtual machine.
+* `NodePort` and `LoadBalancer` services enable connection
```

Comment:
If you're just verifying that you can connect from outside the cluster, I'd make that clearer here.

Also, does this depend on whether you actually configured it that way in the service definition? I was unclear above where it says to replace the ClusterIP setting with NodePort or LoadBalancer. Would you create a separate service object for ClusterIP and then another one for NodePort or LoadBalancer?

Reply:
Different service types expose themselves _internally, within the cluster_ or _externally, outside of the cluster._ CLUSTERIP is internal. NODEPORT and LOADBALANCER are external. 

The external service types have very specific Web applications, all documented in detail in the upstream. The most common use case our users will encounter is for internal exposure via CLUSTERIP. 

After discussion with Petr, we decided to keep the information about the external service types general, as not all edge cases (and there are many) have been tested by us. We also considered not exposing the service types at all. The steps for editing the YAMLs are the same whatever service type you choose. The only difference is in connecting the VM (and exposing the service), which varies if exposing an external service type. That variance is covered in the last NOTE.